From b0574a88956a431630a001697ea225085317de33 Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Wed, 4 Jun 2014 20:06:44 -0400
Subject: [net] tcp: add tcp_syncookies mode to allow unconditionally generation of syncookies

Message-id: <1401912404-10861-5-git-send-email-fwestpha@redhat.com>
Patchwork-id: 82578
O-Subject: [PATCH 4/4] [RHEL7.1 net] tcp: add tcp_syncookies mode to allow unconditionally generation of syncookies
Bugzilla: 1084439
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hsowa@redhat.com>
RH-Acked-by: Daniel Borkmann <dborkman@redhat.com>

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1084439
Tested: with tcp_syncookies=2, nstat shows mib counter for syn cookies increases, connections
are established.

cherry-pick.

commit 5ad37d5deee1ff7150a2d0602370101de158ad86
Author: Hannes Frederic Sowa <hannes@stressinduktion.org>
Date:   Fri Jul 26 17:43:23 2013 +0200

    tcp: add tcp_syncookies mode to allow unconditionally generation of syncookies

    | If you want to test which effects syncookies have to your
    | network connections you can set this knob to 2 to enable
    | unconditionally generation of syncookies.

    Original idea and first implementation by Eric Dumazet.

    Cc: Florian Westphal <fw@strlen.de>
    Cc: David Miller <davem@davemloft.net>
    Signed-off-by: Eric Dumazet <eric.dumazet@gmail.com>
    Signed-off-by: Hannes Frederic Sowa <hannes@stressinduktion.org>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/Documentation/networking/ip-sysctl.txt b/Documentation/networking/ip-sysctl.txt
index c412372..1997f9a 100644
--- a/Documentation/networking/ip-sysctl.txt
+++ b/Documentation/networking/ip-sysctl.txt
@@ -440,6 +440,10 @@ tcp_syncookies - BOOLEAN
  SYN flood warnings in logs not being really flooded, your server
  is seriously misconfigured.
 
+ If you want to test which effects syncookies have to your
+ network connections you can set this knob to 2 to enable
+ unconditionally generation of syncookies.
+
 tcp_fastopen - INTEGER
  Enable TCP Fast Open feature (draft-ietf-tcpm-fastopen) to send data
  in the opening SYN packet. To use this feature, the client application
diff --git a/net/ipv4/tcp_ipv4.c b/net/ipv4/tcp_ipv4.c
index 10ab69f..9b27962 100644
--- a/net/ipv4/tcp_ipv4.c
+++ b/net/ipv4/tcp_ipv4.c
@@ -890,7 +890,7 @@ bool tcp_syn_flood_action(struct sock *sk,
   NET_INC_STATS_BH(sock_net(sk), LINUX_MIB_TCPREQQFULLDROP);
 
  lopt = inet_csk(sk)->icsk_accept_queue.listen_opt;
- if (!lopt->synflood_warned) {
+ if (!lopt->synflood_warned && sysctl_tcp_syncookies != 2) {
   lopt->synflood_warned = 1;
   pr_info("%s: Possible SYN flooding on port %d. %s.  Check SNMP counters.\n",
    proto, ntohs(tcp_hdr(skb)->dest), msg);
@@ -1470,7 +1470,8 @@ int tcp_v4_conn_request(struct sock *sk, struct sk_buff *skb)
   * limitations, they conserve resources and peer is
   * evidently real one.
   */
- if (inet_csk_reqsk_queue_is_full(sk) && !isn) {
+ if ((sysctl_tcp_syncookies == 2 ||
+      inet_csk_reqsk_queue_is_full(sk)) && !isn) {
   want_cookie = tcp_syn_flood_action(sk, skb, "TCP");
   if (!want_cookie)
    goto drop;
diff --git a/net/ipv6/tcp_ipv6.c b/net/ipv6/tcp_ipv6.c
index 430cecb..d94101e 100644
--- a/net/ipv6/tcp_ipv6.c
+++ b/net/ipv6/tcp_ipv6.c
@@ -964,7 +964,8 @@ static int tcp_v6_conn_request(struct sock *sk, struct sk_buff *skb)
  if (!ipv6_unicast_destination(skb))
   goto drop;
 
- if (inet_csk_reqsk_queue_is_full(sk) && !isn) {
+ if ((sysctl_tcp_syncookies == 2 ||
+      inet_csk_reqsk_queue_is_full(sk)) && !isn) {
   want_cookie = tcp_syn_flood_action(sk, skb, "TCPv6");
   if (!want_cookie)
    goto drop;
-- 
1.7.1